daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala

```diff
@@ -28,95 +28,112 @@ private[preprocessing] final class TransactionPreprocessor(
-            stakeholders @ _,
-            signatories @ _,
-            choiceObservers @ _,
-            children @ _,
-            exerciseResult @ _,
-            key @ _,
-            byKey @ _,
-            version @ _,
-          ) =>
-        val templateId = template
-        val (cmd, newCids) =
-          commandPreprocessor.unsafePreprocessExercise(templateId, coid, choice, chosenVal)
-        (cmd, (localCids | newCids.filterNot(globalCids), globalCids))
-      case Node.NodeFetch(coid, templateId, _, _, _, _, _, _, _) =>
-        val cmd = commandPreprocessor.unsafePreprocessFetch(templateId, coid)
-        (cmd, acc)
-      case Node.NodeLookupByKey(templateId, _, key, _, _) =>
-        val keyValue = unsafeAsValueWithNoContractIds(key.key)
-        val cmd = commandPreprocessor.unsafePreprocessLookupByKey(templateId, keyValue)
-        (cmd, acc)
+      case create: Node.NodeCreate[Cid] =>
+        commandPreprocessor.unsafePreprocessCreate(create.templateId, create.arg)
+
+      case exe: Node.NodeExercises[_, Cid] =>
+        commandPreprocessor.unsafePreprocessExercise(
+          exe.templateId,
+          exe.targetCoid,
+          exe.choiceId,
+          exe.chosenValue,
+        )
+      case fetch: Node.NodeFetch[Cid] =>
+        val cmd = commandPreprocessor.unsafePreprocessFetch(fetch.templateId, fetch.coid)
+        (cmd, Set.empty)
+      case lookup: Node.NodeLookupByKey[Cid] =>
+        val keyValue = unsafeAsValueWithNoContractIds(lookup.key.key)
+        val cmd = commandPreprocessor.unsafePreprocessLookupByKey(lookup.templateId, keyValue)
+        (cmd, Set.empty)
     }
   }
 
+  private[this] case class Acc(
+      globalCids: Set[ContractId],
+      localCids: Set[ContractId],
+      commands: BackStack[speedy.Command],
+  ) {
+    def update(
+        newInputCids: Iterable[ContractId],
+        newLocalCids: Iterable[ContractId],
+        cmd: speedy.Command,
+    ) = Acc(
+      globalCids ++ newInputCids.filterNot(localCids),
+      localCids ++ newLocalCids,
+      commands :+ cmd,
+    )
+  }
+
+  /*
+   * Translate a transaction tree into a sequence of Speedy commands and collect the global CIDs.
+   *
+   * A contract ID `cid` is *local* w.r.t. a node `n`, if either:
+   *  - it is local in a previous node w.r.t. traversal order, or
+   *  - `n` is a create node such that `n.coid == cid`
+   * A contract ID `cid` is *global* in a root node `n`, if:
+   *  - it is
+   * A contract ID `cid` is *global* in a root node `n`, if
+   *  - `cid` is not local w.r.t. `n`, and
+   *  - if `cid` is an input of a `n`, i.e. :
+   *    * `n` is a create node and `cid` appear in the payload of the create contract (`n.arg`)
+   *    * `n` is an exercise node and `cid` appear in the exercise argument (`n.choosenValue`)
+   *    * `n` is an exercise node and `cid` is the ID of the exercise contract (`n.taretCoid`)
+   * A contract ID is *global* w.r.t a transaction `tx` if it is global w.r.t one of the root `tx`.
```

Comment:
Perhaps worth pointing out in the comment that this is only parts of the definition of global cids. In particular, a contract id is also global if it is referenced in the arguments of a global fetched or lookup’d contract. We detect those cases during interpretation so no need to add it here, just asking to clarify the comment.

Reply:
:+1: